src/main/java/core/basesyntax/service/impl/FruitTransactionServiceImpl.java

```diff
@@ -0,0 +1,43 @@
+package core.basesyntax.service.impl;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.service.FruitTransactionService;
+import core.basesyntax.strategy.OperationStrategy;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class FruitTransactionServiceImpl implements FruitTransactionService {
+    private static final FruitTransaction.Operation
+            TOTAL = FruitTransaction.Operation.TOTAL;
+    private final OperationStrategy operationStrategy;
+
+    public FruitTransactionServiceImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void calculate() {
+        Map<String, List<FruitTransaction>> transactionMap = Storage.fruitTransactions.stream()
+                .collect(Collectors.groupingBy(FruitTransaction::getFruit,
+                        Collectors.mapping(fruitTransaction -> fruitTransaction,
+                                Collectors.toList())));
+        for (Map.Entry<String, List<FruitTransaction>> entry : transactionMap.entrySet()) {
+            int result = 0;
+            List<FruitTransaction> fruitTransactions = entry.getValue();
+            for (FruitTransaction fruitTransaction : fruitTransactions) {
+                result = operationStrategy.get(fruitTransaction.getOperation())
+                        .getOperation(result, fruitTransaction.getQuantity());
+            }
+            if (result < 0) {
+                throw new RuntimeException("Balance mustn't be negative");
+            }
+            FruitTransaction fruitTransaction = new FruitTransaction();
+            fruitTransaction.setOperation(TOTAL);
+            fruitTransaction.setFruit(entry.getKey());
+            fruitTransaction.setQuantity(result);
+            Storage.fruitTransactions.add(fruitTransaction);
+        }
+    }
```

Comment:
По-друге: через те, що в тебе Storage це ліст транзакцій - тут у тебе переускладнена логіка вийшла.
Я не уявляю собі Storage як ліст транзакцій взагалі. Це буде якийсь журнал, ніяк не склад для зберігання фруктів.
Скоріш за все таке рішення вийшло через бажання зберігати десь транзакції. Але особливого сенсу в цьому немає. Якщо ми зробимо як я писав у першому пункті - то ми просто цей ліст транзакцій отримаємо із одного методу і закинемо в інший, де і будемо його обробляти як вхідний аргумент.
Тоді в нашому storage потрібно зберігати не їх, а по суті проміжний результат після обробки кожної транзакції. Нас хвилює тільки назва фрукту і кількість.
Тому якщо ми перетворимо наш storage у Map<String, Integer> де ключ - це фрукт, а value - кількість цього фрукту - то весь код цього методу стане максимально простим:
```suggestion
    @Override
    public void processTransactions(List<FruitTransaction> transactions) {
        for (transaction : transactions) {
             operationStrategy.get(transaction.getOperation()).handleTransaction(transaction);
        }
    }
```